core/src/mast/serialization/info.rs

```diff
@@ -0,0 +1,456 @@
+use miden_crypto::hash::rpo::RpoDigest;
+use winter_utils::{ByteReader, ByteWriter, Deserializable, DeserializationError, Serializable};
+
+use crate::mast::{MastForest, MastNode, MastNodeId, MerkleTreeNode};
+
+use super::{basic_block_data_decoder::BasicBlockDataDecoder, DataOffset};
+
+// MAST NODE INFO
+// ===============================================================================================
+
+/// Represents a serialized [`MastNode`], with some data inlined in its [`MastNodeType`].
+///
+/// In the case of [`crate::mast::BasicBlockNode`], all its operation- and decorator-related data is
+/// stored in the serialized [`MastForest`]'s `data` field at offset represented by the `offset`
+/// field. For all other variants of [`MastNode`], the `offset` field is guaranteed to be 0.
+///
+/// The serialized representation of [`MastNodeInfo`] is guaranteed to be fixed width, so that the
+/// nodes stored in the `nodes` table of the serialzied [`MastForest`] can be accessed quickly by
+/// index.
+#[derive(Debug)]
+pub struct MastNodeInfo {
+    ty: MastNodeType,
+    offset: DataOffset,
+    digest: RpoDigest,
+}
```

Comment:
Do we actually need to track the offset here? My understanding is that `offset` is useful only for basic blocks and for everything else it is set to 0. If so, we could probably move the offset into `MastNodeType` to make it look like so:
```rust
#[repr(u8)]
pub enum MastNodeType {
    Join {
        left_child_id: u32,
        right_child_id: u32,
    } = JOIN,
    Split {
        if_branch_id: u32,
        else_branch_id: u32,
    } = SPLIT,
    Loop {
        body_id: u32,
    } = LOOP,
    Block {
        /// Offset of the basic block in the data segment
        offset: u32,
        /// The number of operations and decorators in the basic block
        len: u32,
    } = BLOCK,
    Call {
        callee_id: u32,
    } = CALL,
    SysCall {
        callee_id: u32,
    } = SYSCALL,
    Dyn = DYN,
    External = EXTERNAL,
}
```

Reply:
I think we might still needed. My understanding is that the current code assumes that we are deserializing the entire MAST forest - and in this case, we always know the location of the next basic block data implicitly.

However, I think one of the goals of this format is to make it possible to deserialize parts of the forest (no need to implement this right now, but we may want to have support for this in the format itself). For example, we may start with one of the roots and then "jump around" the file to read only the nodes relevant to that root. We can already do this for all nodes because `NodeType` is fixed size (so, given a node index, we can compute the location of this node without having to read the entire `nodes` data), but without associating offset into the data section for basic blocks, I'm not sure how we'd do it.

Reply:
> This is a happy accident that saves us 4 bytes per `MastNode` 😄.

If we encode the offset into the `MastNodeType` - we would also save on these 4 bytes, right?

Reply:
> However, I think one of the goals of this format is to make it possible to deserialize parts of the forest (no need to implement this right now, but we may want to have support for this in the format itself).

Ah yes you're totally right, I forgot about that.

> If we encode the offset into the MastNodeType - we would also save on these 4 bytes, right?

Absolutely; at the time of writing I somehow forgot that your initial suggestion had the exact same benefit, sorry about that.

I implemented the suggestion. The only nit I'm not happy about with the current implementation is that we currently ignore the offset stored in the `Block`, and use the one implicit in the `SliceReader`. However, someone could technically move blocks around during serialization, such that they're not in the order we expect. So at the very least, we should confirm that the block's offset corresponds to our `SliceReader`'s offset, which would require adding a method `SliceReader::position(&self)` to `SliceReader`.

Reply:
> The only nit I'm not happy about with the current implementation is that we currently ignore the offset stored in the `Block`, and use the one implicit in the `SliceReader`.

I described a potential solution for this in https://github.com/0xPolygonMiden/miden-vm/pull/1370#discussion_r1674755804.